src/state/utils/store_utils.ts

```diff
@@ -13,15 +13,15 @@ import { SubscribableStore } from "../types";
  * @param listenerFn The listener function that computes and returns the new
  * derived value, called when the selected dependencies change.
  */
-export const addDerivedStateSubscriber = <T, const U = [keyof T][]>(
+export const addDerivedStateSubscriber = <T, const U>(
   store: SubscribableStore<T>,
   selectorFn: (state: T) => U,
-  listenerFn: (state: U) => Partial<T>
+  listenerFn: (state: U, prevState: U) => Partial<T>
 ): void => {
   store.subscribe(
     selectorFn,
-    (selectedState) => {
-      store.setState(listenerFn(selectedState));
+    (selectedState, prevSelectedState) => {
+      store.setState(listenerFn(selectedState, prevSelectedState));
```

Comment:
The selected state can be anything, including an array or an object. I updated the types to match, and also added the optional `prevState` param.